lib/phoenix/router.ex

```diff
@@ -464,75 +464,50 @@ defmodule Phoenix.Router do
       is automatically derived from the controller name, i.e. `UserController` will
       have name `"user"`
     * `:as` - configures the named helper exclusively
+    * `:singleton` - Defines "RESTful" routes for a resource that client's lookup without referencing an ID, which will include the following routes instead:
 
-  """
-  defmacro resources(path, controller, opts, do: nested_context) do
-    add_resources path, controller, opts, do: nested_context
-  end
+    * `GET /user` => `:show`
+    * `GET /user/new` => `:new`
+    * `POST /user` => `:create`
+    * `GET /user/edit` => `:edit`
+    * `PATCH /user` => `:update`
+    * `PUT /user` => `:update`
+    * `DELETE /user` => `:delete`
 
-  @doc """
-  See `resources/4`.
-  """
-  defmacro resources(path, controller, do: nested_context) do
-    add_resources path, controller, [], do: nested_context
-  end
 
-  defmacro resources(path, controller, opts) do
-    add_resources path, controller, opts, do: nil
-  end
 
-  @doc """
-  See `resources/4`.
   """
-  defmacro resources(path, controller) do
-    add_resources path, controller, [], do: nil
-  end
-
-  @doc """
-  Defines "RESTful" routes for a resource that client's lookup without referencing an ID.
-
-  The given definition:
-
-      resource "/account", UserController
-
-  will include routes to the following actions:
-
-    * `GET /account` => `:show`
-    * `GET /account/new` => `:new`
-    * `POST /account` => `:create`
-    * `GET /account/edit` => `:edit`
-    * `PATCH /account` => `:update`
-    * `PUT /account` => `:update`
-    * `DELETE /account` => `:delete`
-
-  ## Options
-
-  This macro accepts the same options as `resources/4`
-
-  """
-  defmacro resource(path, controller, opts, do: nested_context) do
-    add_resource path, controller, opts, do: nested_context
+  defmacro resources(path, controller, opts, do: nested_context) do
+    if opts[:singleton] do
```

Comment:
Unfortunately this won't work because you are expecting `opts[:singleton]` to be available at compile time. So if someone calls as:

    opts = [singleton: true]
    resources "/users", UserController, opts

It wouldn't work. We need to move it to inside the functions. We should probably rename `Resource.plural` and `Resource.singular` to a single function called `Resource.build` and then choose which functions to invoke.